File: backend/src/main/java/com/devops/backend/service/ResourceMonitoringService.java
package com.devops.backend.service;

import com.devops.backend.dto.monitoring.FluctuationPoint;
import com.devops.backend.dto.monitoring.ResourceHierarchyNode;

import java.time.Instant;
import java.util.List;

/**
 * Read model for the Resource Monitoring analytics section.
 * Visible to both USER and DEVOPS roles — pricing data is filtered out
 * for USER callers upstream in the controller.
 */
public interface ResourceMonitoringService {

    /** Full Environment → Cluster → Project → Microservice tree. */
    List<ResourceHierarchyNode> getHierarchy();

    /** Subtree rooted at a single environment (for lazy loading). */
    List<ResourceHierarchyNode> getEnvironmentSubtree(String environment);

    /** Fluctuation timeline at an environment/cluster scope. */
    List<FluctuationPoint> getClusterFluctuation(String environment, String clusterName,
                                                 Instant from, Instant to);

    /** Fluctuation timeline for a project (summed across microservices). */
    List<FluctuationPoint> getProjectFluctuation(String projectId, String environment,
                                                 Instant from, Instant to);

    /** Fluctuation timeline for a single microservice. */
    List<FluctuationPoint> getMicroserviceFluctuation(String microserviceId,
                                                      Instant from, Instant to);

    /** Capture a snapshot of a project's current config (manual trigger for DevOps). */
    int snapshotProject(String projectId, String environment, String actor);

    /** Capture snapshots for every project & environment — invoked by scheduler. */
    int snapshotAll(String source);
}

File: frontend/src - Copy/services/resourceMonitoringService.js
import { apiRequest } from "./apiClient";

const q = (obj) => {
    const p = new URLSearchParams();
    Object.entries(obj || {}).forEach(([k, v]) => {
        if (v !== undefined && v !== null && v !== "") p.set(k, String(v));
    });
    const s = p.toString();
    return s ? `?${s}` : "";
};

/** Environment → Cluster → Project → Microservice tree. */
export const getResourceHierarchy = (environment) =>
    apiRequest(`/resource-monitoring/hierarchy${q({ environment })}`);

export const getClusterFluctuation = ({ environment, clusterName, from, to }) =>
    apiRequest(`/resource-monitoring/fluctuation/cluster${q({ environment, clusterName, from, to })}`);

export const getProjectFluctuation = ({ projectId, environment, from, to }) =>
    apiRequest(`/resource-monitoring/fluctuation/project${q({ projectId, environment, from, to })}`);

export const getMicroserviceFluctuation = ({ microserviceId, from, to }) =>
    apiRequest(`/resource-monitoring/fluctuation/microservice${q({ microserviceId, from, to })}`);

/** DevOps/Admin only — snapshot current config for a project/environment. */
export const snapshotProject = ({ projectId, environment }) =>
    apiRequest("/resource-monitoring/snapshot", {
        method: "POST",
        body: JSON.stringify({ projectId, environment }),
    });
